fendermint/app/src/app.rs

```diff
@@ -759,42 +759,60 @@ where
     async fn commit(&self) -> AbciResult<response::Commit> {
         let exec_state = self.take_exec_state().await;
 
-        // Commit the execution state to the datastore.
-        let mut state = self.committed_state()?;
-        state.block_height = exec_state.block_height().try_into()?;
-        state.state_params.timestamp = exec_state.timestamp();
+        let block_timestamp = exec_state.timestamp();
+        let block_height: BlockHeight = exec_state.block_height().try_into()?;
 
+        // Commit the execution state to the datastore.
         let (
             state_root,
             FvmUpdatableParams {
                 power_scale,
                 circ_supply,
             },
-            _,
+            is_dirty,
         ) = exec_state.commit().context("failed to commit FVM")?;
 
-        state.state_params.state_root = state_root;
-        state.state_params.power_scale = power_scale;
-        state.state_params.circ_supply = circ_supply;
+        // Update the application state.
+        let mut state = self.committed_state()?;
 
-        let app_hash = state.app_hash();
-        let block_height = state.block_height;
+        // CometBFT decided to create a block, so we must remember what height we're at,
+        // but we can opt to not change the application hash if there was no meaningful change.
+        state.block_height = block_height;
+
+        // If there has been no update to the state then we can skip updating the the timestamp,
+        // because it would be the only thing changing the application state, which would cause
+        // CometBFT to produce an empty block even if it's configured to skip them.
+        // Leaving the timestamp unchanged only affects historical queries targetting these heights.
+        // The state timestamp will catch up with the header timestamp as soon as there is a block
+        // which does something to the ledger. This could be an empty block causing a checkpoint.
+        // See https://docs.cometbft.com/v0.37/core/configuration#empty-blocks-vs-no-empty-blocks
+        let update_params = is_dirty || state_root != state.state_params.state_root;
```

Comment:
I played with this and it seems that in every call to `commit`, `state_root != state.state_params.state_root`, even if there are no txns. 

I can confirm that if you force `update_params` to `false`, comet stops calling `commit` when `--consensus.create_empty_blocks=false` is used.

Reply:
Thanks for trying! 

Ah, I remember now, that's because we have this new step where the block hash is [inserted into the ledger](https://github.com/consensus-shipyard/ipc/blob/c6c5142ddbbcd8b24bbcffc5bbd79faa4aee89c0/fendermint/vm/interpreter/src/fvm/exec.rs#L85-L113) so that we can implement the `BLOCKHASH` function in the `evm` actor.

Hm, I can't see how we could get around that causing a new empty block from being created :thinking: 

The only way would be if we didn't save the hash of empty blocks, but I don't think breaking the 1-to-1 correspondence of the API and the ledger would be a good idea.

Reply:
I added a setting to disable chain metadata, in case you can live without it in your tests.

Reply:
Ah! Makes sense. The setting looks great. Thanks for the quick help on this :pray: